posthog/temporal/data_imports/external_data_job.py

```diff
@@ -0,0 +1,219 @@
+
+
+@activity.defn
+async def validate_schema_activity(inputs: ValidateSchemaInputs) -> bool:
+    return await sync_to_async(is_schema_valid)(  # type: ignore
+        source_schemas=inputs.source_schemas,
+        external_data_source_id=inputs.external_data_source_id,
+        create=inputs.create,
+    )
+
+
+@dataclasses.dataclass
+class MoveDraftToProductionExternalDataJobInputs:
+    team_id: int
+    external_data_source_id: str
+
+
+@activity.defn
+async def move_draft_to_production_activity(inputs: MoveDraftToProductionExternalDataJobInputs) -> None:
+    await sync_to_async(move_draft_to_production)(  # type: ignore
+        team_id=inputs.team_id,
+        external_data_source_id=inputs.external_data_source_id,
+    )
+
+
+@dataclasses.dataclass
+class ExternalDataJobInputs:
+    team_id: int
+    external_data_source_id: str
+
+
+@activity.defn
+async def run_external_data_job(inputs: ExternalDataJobInputs) -> List[SourceSchema]:
+    model: ExternalDataSource = await sync_to_async(get_external_data_source)(  # type: ignore
+        team_id=inputs.team_id,
+        external_data_source_id=inputs.external_data_source_id,
+    )
+
+    job_inputs = PIPELINE_TYPE_INPUTS_MAPPING[model.source_type](
+        team_id=inputs.team_id, job_type=model.source_type, dataset_name=model.draft_folder_path, **model.job_inputs
+    )
+    job_fn = PIPELINE_TYPE_RUN_MAPPING[model.source_type]
+
+    heartbeat_details = AsyncHeartbeatDetails()
+    func = heartbeat_details.make_activity_heartbeat_while_running(job_fn, dt.timedelta(seconds=10))
+
+    return await func(job_inputs)
+
+
+# TODO: update retry policies
+@workflow.defn(name="external-data-job")
+class ExternalDataJobWorkflow(PostHogWorkflow):
+    @staticmethod
+    def parse_inputs(inputs: list[str]) -> ExternalDataJobInputs:
+        loaded = json.loads(inputs[0])
+        return ExternalDataJobInputs(**loaded)
+
+    @workflow.run
+    async def run(self, inputs: ExternalDataJobInputs):
+        # create external data job and trigger activity
+        create_external_data_job_inputs = CreateExternalDataJobInputs(
+            team_id=inputs.team_id,
+            external_data_source_id=inputs.external_data_source_id,
+        )
+
+        run_id = await workflow.execute_activity(
+            create_external_data_job_model,
+            create_external_data_job_inputs,
+            start_to_close_timeout=dt.timedelta(minutes=1),
+            retry_policy=RetryPolicy(
+                initial_interval=dt.timedelta(seconds=10),
+                maximum_interval=dt.timedelta(seconds=60),
+                maximum_attempts=0,
+                non_retryable_error_types=["NotNullViolation", "IntegrityError"],
+            ),
+        )
+
+        update_inputs = UpdateExternalDataJobStatusInputs(
+            id=run_id, run_id=run_id, status=ExternalDataJob.Status.COMPLETED, latest_error=None
+        )
+
+        try:
+            # TODO: can make this a child workflow for separate worker pool
```

Comment:
question: Couldn't all of these activities be one activity? More activities means we pay more for Temporal cloud, and it would appear `validate_schema_activity`,  `move_draft_to_production_activity`, are activities that just call a single function. Moreover, they have similar (or equal) retry policies.

Reply:
Ultimately up to you, I'm still learning how to adopt Temporal concepts into code. And while volume is low, the costs will not be high anyways.

Reply:
probably a premature optimization on my part here. I was mainly thinking for the data import job, if it became a long running activity, we could have it run on a separate pool of workers. For the rest of them, yes I think we can run them all in one but as far as I understood from my crash course on temporal is that each activity should be an isolated function call as possible

Reply:
If you are not blocking the workers, I don't think you would need a separate pool (and I would  _strongly_ recommend that we don't block). So, I would argue the optimization work should be target not blocking the current pool instead of sending tasks to a new pool of workers. 

> each activity should be an isolated function call as possible

Ideally, each activity should represent an idempotent action. However, said action can involve multiple steps (i.e. function calls). I think the idempotency of the whole action is the more important property that we should strive to maintain when deciding whether to break apart activities or not. That being said, even then there is a lot of margin, so I tend to also look at retry policies and inputs to decide whether to combine activites or not.

Reply:
Fine to leave this as it is by the way, we are not experts in Temporal yet and we can learn stuff as we go :+1: